File: Python/_030_Substring_with_Concatenation_of_All_Words/solution.py
class Solution:
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if len(words) == 0:
            return []
        word_count = {}
        word_count_temp = {}
        res = []
        for i in range(len(words)):
            if words[i] in word_count:
                word_count[words[i]] += 1
            else:
                word_count[words[i]] = 1
        for i in range(0, len(s) - len(words[0]) * len(words) + 1):
            j = 0
            word_count_temp.clear()
            while j < len(words):
                word = s[i + j * len(words[0]) : i + j * len(words[0]) + len(words[0])]
                if word in word_count:
                    if word in word_count_temp:
                        word_count_temp[word] += 1
                    else:
                        word_count_temp[word] = 1
                    if word_count_temp[word] > word_count[word]:
                        break
                else:
                    break
                j += 1
            if j == len(words):
                res.append(i)
        return res

File: Java/_030_Substring_with_Concatenation_of_All_Words/Solution.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        if (words.length == 0)
            return new ArrayList<Integer>();   
        int word_size = words[0].length();
        int word_nums = words.length;
        Map<String, Integer> word_count = new HashMap<String, Integer>();
        Map<String, Integer> word_count_temp = new HashMap<String, Integer>();
        List<Integer> res = new ArrayList<Integer>();
        for (int i = 0; i < word_nums; i++) {
            if (word_count.containsKey(words[i])) {
                int count = word_count.get(words[i]) + 1;
                word_count.put(words[i], count);
            }
            else {
                word_count.put(words[i], 1);
            }
        }
        for (int i = 0; i < s.length() - word_nums * word_size + 1; i++) {
            int j = 0;
            word_count_temp.clear();
            for (j = 0; j < word_nums; j++) {
                String word = s.substring(i + j * word_size, i + j * word_size + word_size);
                if (word_count.containsKey(word)) {
                    if (word_count_temp.containsKey(word)) {
                        int count = word_count_temp.get(word) + 1;
                        word_count_temp.put(word, count);
                    }
                    else {
                        word_count_temp.put(word, 1);
                    }
                    if (word_count_temp.get(word) > word_count.get(word))
                        break;
                }
                else
                    break;
            }
            if (j == word_nums)
                res.add(i);
        }
        return res;
    }
}
